embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/arch/DataSourceState.kt

```diff
@@ -0,0 +1,68 @@
+package io.embrace.android.embracesdk.arch
+
+/**
+ * Holds the current state of the service. This class automatically handles changes in config
+ * that enable/disable the service, and creates new instances of the service as required.
+ * It also is capable of disabling the service if the [EnvelopeType] is not supported.
+ */
+internal class DataSourceState<T : DataSource<R>, R>(
+
+    /**
+     * Provides instances of services. A service must define an interface
+     * that extends [DataSource] for orchestration. This helps enforce testability
+     * by making it impossible to register data capture without defining a testable interface.
+     */
+    factory: () -> DataSource<R>,
+
+    /**
+     * Predicate that determines if the service should be enabled or not, via a config value.
+     */
+    private val configGate: () -> Boolean,
```

Comment:
Are we able to pass in config settings via a similar interface? Perhaps the return value can be a generic object of some sort?

Reply:
I'm not sure what this means? My current idea is that we would supply a boolean value from `ConfigService` as a parameter, e.g.:

```
class DataSourceState(
    configGate = { configService.dataCaptureBehavior.isAnrCaptureEnabled() }
...
```

Happy to iterate if you have other ideas